File: Ash/OS_LAB/14_lrupr.c
#include <stdio.h>

void lruPageReplacement(int pages[], int n, int capacity)
{
    int frame[capacity];
    int pageFaults = 0;
    int recentCount = 0;
    int recent[capacity];

    for (int i = 0; i < capacity; i++)
    {
        frame[i] = -1;
        recent[i] = 0;
    }

    printf("Incoming\tFrame\n");

    for (int i = 0; i < n; i++)
    {
        int page = pages[i];
        int pageFound = 0;

        // Check if the page is already present in the frame
        for (int j = 0; j < capacity; j++)
        {
            if (frame[j] == page)
            {
                pageFound = 1;
                recent[j] = recentCount++;
                break;
            }
        }

        if (!pageFound)
        {
            // Page is not present, find the least recently used page in the frame
            int lruIndex = 0;
            for (int j = 1; j < capacity; j++)
            {
                if (recent[j] < recent[lruIndex])
                {
                    lruIndex = j;
                }
            }

            // Replace the least recently used page with the new page
            frame[lruIndex] = page;
            recent[lruIndex] = recentCount++;
            pageFaults++;
        }

        // Print the current page and frame status
        printf("%d\t\t", page);
        for (int j = 0; j < capacity; j++)
        {
            if (frame[j] != -1)
                printf("%d\t\t", frame[j]);
            else
                printf("-\t\t");
        }
        printf("\n");
    }

    printf("\nTotal Page Faults: %d\n", pageFaults);
}

int main()
{
    int n;

    // Input the number of pages
    printf("Enter the number of pages: ");
    scanf("%d", &n);

    int pages[n];

    // Input the sequence of page requests
    printf("Enter the sequence of page requests:\n");
    for (int i = 0; i < n; i++)
        scanf("%d", &pages[i]);

    int capacity;

    // Input the number of frames
    printf("Enter the number of frames: ");
    scanf("%d", &capacity);

    lruPageReplacement(pages, n, capacity);

    return 0;
}

File: Ash/OS_LAB/14_lrupr.py
def lru_page_replacement(pages, capacity):
    frame = [-1] * capacity
    page_faults = 0
    recent_count = 0
    recent = [0] * capacity

    print("Incoming\tFrame")

    for page in pages:
        page_found = False

        # Check if the page is already present in the frame
        for j in range(capacity):
            if frame[j] == page:
                page_found = True
                recent[j] = recent_count
                recent_count += 1
                break

        if not page_found:
            # Page is not present, find the least recently used page in the frame
            lru_index = 0
            for j in range(1, capacity):
                if recent[j] < recent[lru_index]:
                    lru_index = j

            # Replace the least recently used page with the new page
            frame[lru_index] = page
            recent[lru_index] = recent_count
            recent_count += 1
            page_faults += 1

        # Print the current page and frame status
        print(f"{page}\t\t", end="")
        for j in range(capacity):
            if frame[j] != -1:
                print(f"{frame[j]}\t\t", end="")
            else:
                print("-\t\t", end="")
        print()

    print(f"\nTotal Page Faults: {page_faults}")

if __name__ == "__main__":
    n = int(input("Enter the number of pages: "))

    # Input the sequence of page requests
    pages = []
    print("Enter the sequence of page requests:")
    for _ in range(n):
        pages.append(int(input()))

    capacity = int(input("Enter the number of frames: "))

    lru_page_replacement(pages, capacity)


